### ISSUE TITLE
podman save does not have an option to produce the same output every time

### ISSUE THREAD

**Issue by graywolf-at-work** [containers/podman#14978]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

Podman save does not produce the same tarball every time it is executed. That
complicated doing reproducible builds.

```
+$ echo 'from scratch' | podman build -f - .
STEP 1/1: FROM scratch
COMMIT
--> 115e55eb5ca
115e55eb5ca5fc95deb6532fc83cebaac7ad89684719a2538776f17ee884c471
+$ podman save --format oci-archive -o a.tar 115e55eb5ca5fc95deb6532fc83cebaac7ad89684719a2538776f17ee884c471
Copying blob 5f70bf18a086 done
Copying config 115e55eb5c done
Writing manifest to image destination
Storing signatures
:$ podman save --format oci-archive -o b.tar 115e55eb5ca5fc95deb6532fc83cebaac7ad89684719a2538776f17ee884c471
Copying blob 5f70bf18a086 done
Copying config 115e55eb5c done
Writing manifest to image destination
Storing signatures
+$ sha256sum *.tar
3aa779c6c0aa936ee1edac3e49527477ebfafe4c2c6d76f020ff918cf906923a  a.tar
14f3a33901726efab7e953ef8ab04b71895b36e0c561b7ffb6416acd97943449  b.tar
```

**Describe the results you received:**

The two tarball are not the same.

**Describe the results you expected:**

The two tarball are the same.

**Additional information you deem important (e.g. issue happens only occasionally):**

Current workaround I have is this repackaging script:

```
#!/bin/sh
set -eu

tmpd=/tmp/repackage-tar.$$

cleanup() {
        rc=$?
        set +e

        rm -rf $v "$tmpd"

        exit "$rc"
}
trap cleanup EXIT

f=$1; shift

mkdir "$tmpd"
tar -tf "$f" | sort >"$tmpd/before"

if grep -q ^/ <"$tmpd/before"; then
        echo >&2 'Some records have absolute paths'
        exit 1
fi

mkdir "$tmpd/data"
tar -C "$tmpd/data" -xf "$f"
tar -C "$tmpd/data" --sort=name --mtime=@0 --owner=0 --group=0 --numeric-owner \
    --pax-option=exthdr.name=%d/PaxHeaders/%f,delete=atime,delete=ctime \
    -cf "$f" --no-recursion --verbatim-files-from -T "$tmpd/before"
tar -tf "$f" >"$tmpd/after"

diff -u "$tmpd/before" "$tmpd/after"
```

So there is a way around this problem, but it definitely is suboptimal that
this is necessary. Ideal solution would be to have something like
`--reproducible` flag for `podman save` that would make sure to always produce
bit-to-bit equal tarballs given specific image id.

**Output of `podman version`:**

```
+$ podman version
Client:       Podman Engine
Version:      4.1.0
API Version:  4.1.0
Go Version:   go1.18.2
Git Commit:   73c021e0588393f18d2c54dec335ec9e960287a4
Built:        Thu May 12 02:00:55 2022
OS/Arch:      linux/amd64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

No

**Comment by flouthoc** [comment]
Thanks for reporting at first instance looks like issue with `c/image` especially while we perform commit of archive.

**Comment by flouthoc** [comment]
@vrothberg @mtrmac Correct me if i got this wrong, I think `c/image/oci` first unpacks the dest into a new temporary dir and then packs it again to a tar, this process can cause change of timestamps in between causing final `sha` to be different everytime https://github.com/containers/image/blob/main/oci/archive/oci_dest.go#L159

**Comment by vrothberg** [comment]
Yes, it is the timestamp that differ.  When unpacking the tar's, the contents are identical.

**Comment by mtrmac** [comment]
Strictly speaking, I’m skeptical that much of the code makes any reproducibility guarantees. Notably `json.Marshal` says nothing on the topic.

That said, a c/image option to pass a specific time (similar to https://wiki.debian.org/ReproducibleBuilds/TimestampsInTarball ) seems reasonable. I don’t immediately have an opinion on what the default should be.

**Comment by rhatdan** [comment]
Default should be EPOCH

**Comment by rhatdan** [comment]
I should say the default should be current if users want to override it they can set to EPOCH.

**Comment by rhatdan** [comment]
@mtrmac @vrothberg thoughts?

**Comment by micah** [comment]
The reproducible community seems to be settling on having the environment variable SOURCE_DATE_EPOCH as something that can be set to a value, that is then used for all date-specific operations so that the dates stay the same with the same builds.

This appears to be possible with BuildKit (https://medium.com/nttlabs/bit-for-bit-reproducible-builds-with-dockerfile-7cc2b9faed9f).

I have a similar problem, I can feed the exact same .tar to `podman import` and the resulting digests are different. However, I do notice that the "blob" digest stays the same, but the "config" is different, for example:

```
$ cat build.tar | podman import - build:one
Getting image source signatures
Copying blob 82ce02f764d7 done  
Copying config f249c34e29 done  
Writing manifest to image destination
Storing signatures
sha256:f249c34e29498deea8af78d52cf863ab919550805014bc038f71a29845b64b30
 
$ cat build.tar | podman import - build:two
Getting image source signatures
Copying blob 82ce02f764d7 skipped: already exists  
Copying config 4e63c9cf3b done  
Writing manifest to image destination
Storing signatures
sha256:4e63c9cf3b1914f6dcf1358fb1aa141439aa03e8ce6c936fad989b8eb2a59648
$
```

**Comment by rhatdan** [comment]
If you are interested in opening a PR I am sure it would be considered. podman build currently has an option like this, I believe.

podman build --help | grep timestamp
      --timestamp int                                set created timestamp to the specified epoch seconds to allow for deterministic builds, defaults to current time
